torchtune/training/checkpointing/_utils.py

```diff
@@ -277,3 +299,161 @@ def update_state_dict_for_classifier(
+    """
+
+    for root, dirs, files in os.walk(input_dir):
+
+        # Filter out directories that start with '.'. E.g. ".cache/"
+        dirs[:] = [d for d in dirs if not d.startswith(".")]
+
+        # Construct the corresponding directory in the output
+        relative_path = os.path.relpath(root, input_dir)
+        dest_dir = os.path.join(output_dir, relative_path)
+
+        # Create the directory in the output if it doesn't exist
+        os.makedirs(dest_dir, exist_ok=True)
+
+        for file in files:
+            # Skip files that start with '.'. E.g. ".git"
+            if file.startswith("."):
+                continue
+
+            # Check if the file has one of the specified suffixes
+            if ignore_suffixes and any(
+                file.endswith(suffix) for suffix in ignore_suffixes
+            ):
+                continue
+
+            src_file = os.path.join(root, file)
+            dest_file = os.path.join(dest_dir, file)
+
+            # Copy the file if it doesn't already exist in the destination
+            if not os.path.exists(dest_file):
+                shutil.copy2(src_file, dest_file)
+
+    return
+
+
+def get_recipe_checkpoint_path(
+    output_dir: Path,
+    recipe_checkpoint: Optional[str] = None,
+    resume_from_checkpoint: bool = False,
+):
+    """
+    If recipe_checkpoint is None, look for recipe_state.pt in {output_dir}/{RECIPE_STATE_DIRNAME}/recipe_state.pt
+    This is to make it easier to resume from a previous run, without having to specify the recipe_checkpoint.
+
+    Args:
+        output_dir (Path): Directory containing the recipe checkpoint.
+        recipe_checkpoint (Optional[str]): Name of the recipe checkpoint file. Defaults to None.
+        resume_from_checkpoint (bool): Whether to resume from a checkpoint.
+    Returns:
+        Path: Path to the recipe checkpoint file.
+    Raises:
+        ValueError: If resume_from_checkpoint is True and the recipe checkpoint file is missing.
+    """
+    if not resume_from_checkpoint:
+        return None
+
+    if recipe_checkpoint:
+        recipe_checkpoint = os.path.join(output_dir, recipe_checkpoint)
+    else:
+        # Look for the recipe checkpoint in the output directory
+        tentative_recipe_state_path = os.path.join(
+            output_dir,
+            RECIPE_STATE_DIRNAME,
+            "recipe_state.pt",
+        )
+        if os.path.exists(tentative_recipe_state_path):
+            recipe_checkpoint = tentative_recipe_state_path
+        else:
+            raise ValueError(
+                "If resume_from_checkpoint is True, recipe_checkpoint file must be provided "
+                f"or exist at {tentative_recipe_state_path}."
+            )
+
+    return Path(recipe_checkpoint)
+
+
+def get_adapter_checkpoint_path(
+    output_dir: Path,
+    adapter_checkpoint: Optional[str] = None,
+    resume_from_checkpoint: bool = False,
+    pattern: str = r"^epoch_(\d+)",
+):
```

Comment:
add return types to these (e.g. here we return optional path but for recipe checkpoint we always return path)